### PR TITLE
Add provider and provider_unique_id fields to HUser

### PR DESCRIPTION
**Pull request by seanh**
`LTILaunchResource` currently has three separate properties:

1. `LTILaunchResource.h_user`. Returns a `models.HUser` object.

2. `LTILaunchResource.h_provider`

3. `LTILaunchResource.h_provider_unique_id`

`h_provider` and `h_provider_unique_id` are really part of the `HUser`:
they're only really used by `LTILaunchResource.h_user`.

Add two new attributes `provider` and `provider_unique_id` to `models.HUser`,
change `LTILaunchResource.h_user` to set them, and delete
`LTILaunchResource.h_provider` and `LTILaunchResource.h_provider_unique_id`.

Code that was previously reading `context.h_provider` and
`context.h_provider_unique_id` can now read `context.h_user.provider`
and `context.h_user.provider_unique_id` instead.

This also simplifies the `HAPI` service as `HAPI.create_user()` and
`HAPI.upsert_user()` no longer need to take a separate `provider` and
`provider_unique_id` params in addition to the `h_user` param.

`provider` and `provider_unique_id` are added to as optional fields to `HUser` because there are a couple of places other than `LTILaunchResource` that create `HUser` objects and setting a `provider` and `provider_unique_id` doesn't make sense in those places.

### CODE REVIEW

File: lms/resources/lti_launch.py [hunk 1]
```diff
@@ -31,17 +31,25 @@ def __init__(self, request):
...
+        provider_unique_id = self._request.lti_user.user_id
+
         def username():
             """Return the h username for the current request."""
             username_hash_object = hashlib.sha1()
-            username_hash_object.update(self.h_provider.encode())
-            username_hash_object.update(self.h_provider_unique_id.encode())
+            username_hash_object.update(provider.encode())
+            username_hash_object.update(provider_unique_id.encode())
             return username_hash_object.hexdigest()[:30]
```
Comment (jon-betts) [inline lms/resources/lti_launch.py:45]: Is this moving into the user object in a different PR?

File: lms/services/h_api.py [hunk 2]
```diff
@@ -42,21 +42,22 @@ def get_user(self, username):
...
 
-    def create_user(self, h_user, provider, provider_unique_id):
+    def create_user(self, h_user):
         """
         Create a user in H.
 
         :arg h_user: the user to be created in h
         :type h_user: HUser
-        :param provider: the "provider" string to send to h
-        :param provider_unique_id:  the "provider_unique_id" string to send to h
```
Comment (jon-betts) [inline lms/services/h_api.py:52]: Nice!

File: lms/services/h_api.py [hunk 3]
```diff
@@ -42,21 +42,22 @@ def get_user(self, username):
...
         """
         user_data = {
             "username": h_user.username,
             "display_name": h_user.display_name,
             "authority": self._authority,
```
Comment (jon-betts) [inline lms/services/h_api.py:55]:
```suggestion
            "authority": h_user.authority,
```

I think this works, and it would make it clearer that this is all about the user. Might pave the way for removing `_authority` from this object in future if it's not used.

File: lms/services/h_api.py [hunk 4]
```diff
@@ -77,7 +78,7 @@ def update_user(self, h_user):
             data={"display_name": h_user.display_name},
         )
```
Comment (jon-betts) [inline lms/services/h_api.py:79]: ~I think in future it would be totally Ok for the HUser to know how to turn itself into an H compatible dict (`as_dict()`?) as all the parts come from it.~

Yeah ignore that. This is the H-API. While this is in charge of making API calls, we don't want to have to go hunting around for what the payload is in other files.

This might be more appropriate when the Bulk API is in place.

File: lms/services/h_api.py [hunk 5]
```diff
@@ -77,7 +78,7 @@ def update_user(self, h_user):
...
 
-    def upsert_user(self, h_user, provider, provider_unique_id):
+    def upsert_user(self, h_user):
```
Comment (jon-betts) [inline lms/services/h_api.py:81]: A much nicer interface again

File: lms/services/h_api.py [hunk 6]
```diff
@@ -86,13 +87,11 @@ def upsert_user(self, h_user, provider, provider_unique_id):
...
-        :param provider: the "provider" string to send to h
-        :param provider_unique_id:  the "provider_unique_id" string to send to h
         """
         try:
             self.update_user(h_user)
         except HAPINotFoundError:
-            self.create_user(h_user, provider, provider_unique_id)
+            self.create_user(h_user)
 
     def upsert_group(self, group_id, group_name):
```
Comment (jon-betts) [inline lms/services/h_api.py:96]: A group object like the user would work really well here too.

**Comment by seanh** [comment]
Rebased

File: lms/services/h_api.py (hunk 3 above)
Comment (seanh) [reply lms/services/h_api.py]: Done. We're unlikely to be able to remove `authority` from `HAPIService` though, it's used for a bunch of things

File: lms/services/h_api.py (hunk 4 above)
Comment (seanh) [reply lms/services/h_api.py:79]: namedtuples do come with an `_asdict()` method for free and the resulting dict happens to be close to (but not exactly) what's needed for the h create user API. But not what the update user API needs. It's probably fine for `HAPIService` to handle serializing things into the necessary formats for the different API calls though, as you say

File: lms/resources/lti_launch.py (hunk 1 above)
Comment (seanh) [reply lms/resources/lti_launch.py:45]: That's coming, yes
